File: app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaRS/comm/DanaRS_Packet_Bolus_Get_Calculation_Information.java
package info.nightscout.androidaps.plugins.PumpDanaRS.comm;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.Config;

import com.cozmo.danar.util.BleCommandUtil;

import info.nightscout.androidaps.plugins.PumpDanaR.DanaRPump;

public class DanaRS_Packet_Bolus_Get_Calculation_Information extends DanaRS_Packet {
    private static Logger log = LoggerFactory.getLogger(DanaRS_Packet_Bolus_Get_Calculation_Information.class);

    public DanaRS_Packet_Bolus_Get_Calculation_Information() {
        super();
        opCode = BleCommandUtil.DANAR_PACKET__OPCODE_BOLUS__GET_CALCULATION_INFORMATION;
    }

    @Override
    public void handleMessage(byte[] data) {
        DanaRPump pump = DanaRPump.getInstance();

        int dataIndex = DATA_START;
        int dataSize = 1;
        int error = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        double currentBG = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        int carbohydrate = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        pump.currentTarget = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        pump.currentCIR = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        pump.currentCF = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        dataIndex += dataSize;
        dataSize = 2;
        pump.iob = byteArrayToInt(getBytes(data, dataIndex, dataSize)) / 100d;

        dataIndex += dataSize;
        dataSize = 1;
        pump.units = byteArrayToInt(getBytes(data, dataIndex, dataSize));

        if (pump.units == DanaRPump.UNITS_MMOL) {
            pump.currentCF = pump.currentCF / 100d;
            pump.currentTarget = pump.currentTarget / 100d;
            currentBG = currentBG / 100d;
        }
        if (Config.logDanaMessageDetail) {
            log.debug("Result: " + error);
            log.debug("Pump units: " + (pump.units == DanaRPump.UNITS_MGDL ? "MGDL" : "MMOL"));
            log.debug("Current BG: " + currentBG);
            log.debug("Carbs: " + carbohydrate);
            log.debug("Current target: " + pump.currentTarget);
            log.debug("Current CIR: " + pump.currentCIR);
            log.debug("Current CF: " + pump.currentCF);
            log.debug("Pump IOB: " + pump.iob);
        }
    }

    @Override
    public String getFriendlyName() {
        return "BOLUS__GET_CALCULATION_INFORMATION";
    }
}

File: pump/danars/src/main/kotlin/app/aaps/pump/danars/comm/DanaRSPacketBolusGetCalculationInformation.kt
package app.aaps.pump.danars.comm

import app.aaps.core.interfaces.logging.AAPSLogger
import app.aaps.core.interfaces.logging.LTag
import app.aaps.pump.dana.DanaPump
import app.aaps.pump.danars.encryption.BleEncryption
import javax.inject.Inject

class DanaRSPacketBolusGetCalculationInformation @Inject constructor(
    private val aapsLogger: AAPSLogger,
    private val danaPump: DanaPump
) : DanaRSPacket() {

    init {
        opCode = BleEncryption.DANAR_PACKET__OPCODE_BOLUS__GET_CALCULATION_INFORMATION
        aapsLogger.debug(LTag.PUMPCOMM, "New message")
    }

    override fun handleMessage(data: ByteArray) {
        var dataIndex = DATA_START
        var dataSize = 1
        val error = byteArrayToInt(getBytes(data, dataIndex, dataSize))
        dataIndex += dataSize
        dataSize = 2
        var currentBG = byteArrayToInt(getBytes(data, dataIndex, dataSize)).toDouble()
        dataIndex += dataSize
        dataSize = 2
        val carbohydrate = byteArrayToInt(getBytes(data, dataIndex, dataSize))
        dataIndex += dataSize
        dataSize = 2
        danaPump.currentTarget = byteArrayToInt(getBytes(data, dataIndex, dataSize)).toDouble()
        dataIndex += dataSize
        dataSize = 2
        danaPump.currentCIR = byteArrayToInt(getBytes(data, dataIndex, dataSize))
        dataIndex += dataSize
        dataSize = 2
        danaPump.currentCF = byteArrayToInt(getBytes(data, dataIndex, dataSize)).toDouble()
        dataIndex += dataSize
        dataSize = 2
        danaPump.iob = byteArrayToInt(getBytes(data, dataIndex, dataSize)) / 100.0
        dataIndex += dataSize
        dataSize = 1
        danaPump.units = byteArrayToInt(getBytes(data, dataIndex, dataSize))
        if (danaPump.units == DanaPump.UNITS_MMOL) {
            danaPump.currentCF = danaPump.currentCF / 100.0
            danaPump.currentTarget = danaPump.currentTarget / 100.0
            currentBG /= 100.0
        }
        if (error != 0) failed = true
        aapsLogger.debug(LTag.PUMPCOMM, "Result: $error")
        aapsLogger.debug(LTag.PUMPCOMM, "Pump units: " + if (danaPump.units == DanaPump.UNITS_MGDL) "MGDL" else "MMOL")
        aapsLogger.debug(LTag.PUMPCOMM, "Current BG: $currentBG")
        aapsLogger.debug(LTag.PUMPCOMM, "Carbs: $carbohydrate")
        aapsLogger.debug(LTag.PUMPCOMM, "Current target: " + danaPump.currentTarget)
        aapsLogger.debug(LTag.PUMPCOMM, "Current CIR: " + danaPump.currentCIR)
        aapsLogger.debug(LTag.PUMPCOMM, "Current CF: " + danaPump.currentCF)
        aapsLogger.debug(LTag.PUMPCOMM, "Pump IOB: " + danaPump.iob)
    }

    override val friendlyName: String = "BOLUS__GET_CALCULATION_INFORMATION"
}